Src/FluentAssertions/Common/MethodInfoExtensions.cs

```diff
@@ -3,27 +3,64 @@
 using System.Linq;
 using System.Linq.Expressions;
 using System.Reflection;
+using System.Runtime.CompilerServices;
 
 namespace FluentAssertions.Common
 {
     public static class MethodInfoExtensions
     {
+        private static readonly Lazy<int> ImplementationOptionsMask =
+            new Lazy<int>(() => Enum.GetValues(typeof(MethodImplOptions)).Cast<int>().Sum(x => x));
+
         internal static bool IsAsync(this MethodInfo methodInfo)
         {
             return methodInfo.GetMatchingAttributes<Attribute>(a => a.GetType().FullName.Equals("System.Runtime.CompilerServices.AsyncStateMachineAttribute")).Any();
         }
 
         internal static IEnumerable<TAttribute> GetMatchingAttributes<TAttribute>(this MemberInfo memberInfo, Expression<Func<TAttribute, bool>> isMatchingAttributePredicate) where TAttribute : Attribute
         {
-            return memberInfo.GetCustomAttributes(
+            var customAttributes = memberInfo.GetCustomAttributes(
                 typeof(TAttribute), false)
                 .Cast<TAttribute>()
+                .ToList();
+
+            var methodBase = memberInfo as MethodBase;
+            if (typeof(TAttribute) == typeof(MethodImplAttribute) && methodBase != null)
+            {
+                var (success, methodImplAttribute) = RecreateMethodImplAttribute(methodBase);
+
+                if (success)
+                {
+                    customAttributes.Add(methodImplAttribute as TAttribute);
+                }
+            }
+
+            return customAttributes
                 .Where(isMatchingAttributePredicate.Compile());
         }
 
         internal static bool IsNonVirtual(this MethodInfo method)
         {
             return !method.IsVirtual || method.IsFinal;
         }
+
+        private static (bool success, MethodImplAttribute attribute) RecreateMethodImplAttribute(MethodBase methodBase)
+        {
+            var implementationFlags = methodBase?.MethodImplementationFlags ?? 0;
```

Comment:
You can remove the null conditional operator now after moving the type cast. 
 If something sends a null in here, let them be punished.

Reply:
Sure.